File: services/py/quant_core/domain/rules/universe_eval.py
"""Evaluate a :class:`UniversePlan` against a list of :class:`StockMeta`.

Pure: no IO, no clock — derived fields (``listed_days``) take ``asof``
from the plan.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Final

from quant_core.domain.types.universe_screen import (
    UniverseCompare,
    UniverseField,
    UniverseLogical,
)
from quant_core.errors import QuantError

if TYPE_CHECKING:
    from datetime import date

    from quant_core.domain.types.stock import StockMeta
    from quant_core.domain.types.universe_screen import UniverseExpr, UniversePlan


_ST_PREFIXES: Final[tuple[str, ...]] = ("ST", "*ST", "S*ST", "SST")


def evaluate_universe(plan: UniversePlan, metas: list[StockMeta]) -> list[StockMeta]:
    """Filter ``metas`` down to those satisfying ``plan.expr``."""
    return [m for m in metas if _eval_expr(plan.expr, m, plan.asof)]


def _eval_expr(expr: UniverseExpr, meta: StockMeta, asof: date) -> bool:
    if isinstance(expr, UniverseLogical):
        if expr.op == "not":
            return not _eval_expr(expr.args[0], meta, asof)
        if expr.op == "and":
            return all(_eval_expr(a, meta, asof) for a in expr.args)
        return any(_eval_expr(a, meta, asof) for a in expr.args)
    if isinstance(expr, UniverseCompare):
        return _eval_compare(expr, meta, asof)
    raise QuantError("EVALUATION_FAILED", f"unhandled universe node: {type(expr).__name__}")


def _eval_compare(node: UniverseCompare, meta: StockMeta, asof: date) -> bool:
    left = _resolve_field(node.left, meta, asof)
    right = node.right.value
    op = node.op
    if op == "contains":
        return isinstance(left, str) and isinstance(right, str) and right in left
    if op == "starts_with":
        return isinstance(left, str) and isinstance(right, str) and left.startswith(right)
    if op == "not_starts_with":
        return isinstance(left, str) and isinstance(right, str) and not left.startswith(right)
    # Numeric / ordered compare
    if op == "eq":
        return left == right
    if op == "neq":
        return left != right
    if op in {"gt", "lt", "gte", "lte"}:
        return _ordered_compare(op, left, right)
    raise QuantError("EVALUATION_FAILED", f"unhandled universe compare op: {op!r}")


def _ordered_compare(op: str, left: object, right: object) -> bool:
    """Date / int / Decimal — defer to native ordering.

    The parser aligns ``left`` and ``right`` to the same scalar type
    (e.g. both Decimal, both date), so the comparisons below are safe at
    runtime. The type system can't see that invariant, so we tag the
    operators.
    """
    if isinstance(left, bool) or isinstance(right, bool):
        raise QuantError("EVALUATION_FAILED", "ordered compare not supported on bool")
    try:
        if op == "gt":
            return bool(left > right)  # type: ignore[operator]
        if op == "lt":
            return bool(left < right)  # type: ignore[operator]
        if op == "gte":
            return bool(left >= right)  # type: ignore[operator]
        return bool(left <= right)  # type: ignore[operator]
    except TypeError as exc:
        raise QuantError(
            "EVALUATION_FAILED",
            f"values not orderable: {type(left).__name__} vs {type(right).__name__}",
        ) from exc


def _resolve_field(field: UniverseField, meta: StockMeta, asof: date) -> object:
    name = field.field
    if name == "code":
        return meta.code
    if name == "name":
        return meta.name
    if name == "industries":
        return meta.industries
    if name == "list_date":
        return meta.list_date
    if name == "float_pct":
        return meta.float_pct
    if name == "is_st":
        return _is_st(meta.name)
    if name == "exchange":
        return _exchange_for_code(meta.code)
    if name == "listed_days":
        return (asof - meta.list_date).days
    raise QuantError("EVALUATION_FAILED", f"unhandled universe field: {name!r}")


def _is_st(name: str) -> bool:
    upper = name.strip().upper()
    return any(upper.startswith(p) for p in _ST_PREFIXES)


def _exchange_for_code(code: str) -> str:
    """Match the prefix table in ``akshare_stock_meta._exchange_for_code``.

    Returns a sentinel ``"unknown"`` if the code shape isn't recognised
    (so eq/neq still compare cleanly without raising).
    """
    if not (code.isdigit() and len(code) == 6):
        return "unknown"
    if code.startswith("920"):
        return "bj"
    if code.startswith(("60", "68", "900")):
        return "sh"
    if code.startswith(("00", "30", "20")):
        return "sz"
    if code.startswith(("4", "8")):
        return "bj"
    return "unknown"


__all__ = ["evaluate_universe"]

File: apps/api/src/modules/screen/domain/pure/universe-eval.ts
/**
 * Pure evaluator for the universe-screen DSL. Port of
 * `services/py/quant_core/domain/rules/universe_eval.py`.
 *
 * Operates on the canonical {@link StockMetaDto} shape, plus an optional
 * `snapshotByCode` map that unlocks the snapshot-derived fields
 * (`mkt_cap`, `pe_ttm`, `ret_*`, `dde_*` …). Callers that don't pass a
 * snapshot map get the legacy meta-only behaviour: any comparison
 * against a snapshot field resolves to `null` left-hand-side and
 * therefore evaluates to `false`, excluding the row.
 */

import {
  QuantError,
  type StockMetaDto,
  type StockSnapshotDto,
  type UniverseExpr,
  type UniversePlanAst,
} from '@quant/shared';

import { D } from '../../../../common/decimal.js';
import { UNIVERSE_SNAPSHOT_FIELD_SET } from './screen-fields.js';

const ST_PREFIXES = ['ST', '*ST', 'S*ST', 'SST'] as const;

/**
 * Map from code → snapshot. The evaluator only reads from it; we type
 * the input as a plain `Map` (not `ReadonlyMap`) for ergonomic callers
 * but never mutate.
 */
export type SnapshotByCode = ReadonlyMap<string, StockSnapshotDto>;

export function evaluateUniverse(
  plan: UniversePlanAst,
  metas: readonly StockMetaDto[],
  snapshotByCode?: SnapshotByCode,
): StockMetaDto[] {
  const asof = parseIsoDate(plan.asof);
  const snaps = snapshotByCode ?? EMPTY_SNAPSHOT_MAP;
  return metas.filter((m) => evalExpr(plan.expr, m, asof, snaps.get(m.code) ?? null));
}

const EMPTY_SNAPSHOT_MAP: SnapshotByCode = new Map<string, StockSnapshotDto>();

function evalExpr(
  expr: UniverseExpr,
  meta: StockMetaDto,
  asof: Date,
  snap: StockSnapshotDto | null,
): boolean {
  if (expr.kind === 'logical') {
    if (expr.op === 'not') {
      const first = expr.args[0];
      if (first === undefined) {
        throw new QuantError('EVALUATION_FAILED', "universe 'not' requires an arg", {});
      }
      return !evalExpr(first, meta, asof, snap);
    }
    if (expr.op === 'and') {
      for (const a of expr.args) if (!evalExpr(a, meta, asof, snap)) return false;
      return true;
    }
    for (const a of expr.args) if (evalExpr(a, meta, asof, snap)) return true;
    return false;
  }
  // expr.kind === 'compare'
  return evalCompare(expr, meta, asof, snap);
}

function evalCompare(
  node: Extract<UniverseExpr, { kind: 'compare' }>,
  meta: StockMetaDto,
  asof: Date,
  snap: StockSnapshotDto | null,
): boolean {
  const left = resolveField(node.left.field, meta, asof, snap);
  const right = node.right.value;
  const op = node.op;
  // Missing snapshot fields surface as null/undefined. Any comparison
  // against a null LHS evaluates false — semantically "exclude this row"
  // rather than "throw" — which matches the storage-layer convention
  // that null = "no data" everywhere else in the pipeline.
  if (left === null || left === undefined) return false;
  if (op === 'contains') {
    return typeof left === 'string' && typeof right === 'string' && left.includes(right);
  }
  if (op === 'starts_with') {
    return typeof left === 'string' && typeof right === 'string' && left.startsWith(right);
  }
  if (op === 'not_starts_with') {
    return typeof left === 'string' && typeof right === 'string' && !left.startsWith(right);
  }
  if (op === 'eq') return scalarEq(left, right);
  if (op === 'neq') return !scalarEq(left, right);
  if (op === 'gt' || op === 'lt' || op === 'gte' || op === 'lte') {
    return orderedCompare(op, left, right);
  }
  throw new QuantError('EVALUATION_FAILED', `unhandled universe compare op: ${op}`, {});
}

function scalarEq(a: unknown, b: unknown): boolean {
  if (a === null || b === null || a === undefined || b === undefined) return a === b;
  if (typeof a === 'boolean' || typeof b === 'boolean') return a === b;
  if (typeof a === 'string' && typeof b === 'string') return a === b;
  if (a instanceof Date && b instanceof Date) return a.getTime() === b.getTime();
  if (a instanceof Date && typeof b === 'string') {
    return a.getTime() === parseIsoDate(b).getTime();
  }
  if (typeof a === 'string' && b instanceof Date) {
    return parseIsoDate(a).getTime() === b.getTime();
  }
  // Numeric — coerce to Decimal for safety.
  try {
    return new D(a as string | number).eq(new D(b as string | number));
  } catch {
    return false;
  }
}

function orderedCompare(op: 'gt' | 'lt' | 'gte' | 'lte', left: unknown, right: unknown): boolean {
  if (typeof left === 'boolean' || typeof right === 'boolean') {
    throw new QuantError('EVALUATION_FAILED', 'ordered compare not supported on bool', {});
  }
  // Dates compare lexically as ISO strings already, or via getTime.
  if (
    left instanceof Date ||
    right instanceof Date ||
    isIsoDateString(left) ||
    isIsoDateString(right)
  ) {
    const l = coerceDate(left);
    const r = coerceDate(right);
    if (l === null || r === null) return false;
    switch (op) {
      case 'gt':
        return l.getTime() > r.getTime();
      case 'lt':
        return l.getTime() < r.getTime();
      case 'gte':
        return l.getTime() >= r.getTime();
      case 'lte':
        return l.getTime() <= r.getTime();
    }
  }
  try {
    const a = new D(left as string | number);
    const b = new D(right as string | number);
    switch (op) {
      case 'gt':
        return a.gt(b);
      case 'lt':
        return a.lt(b);
      case 'gte':
        return a.gte(b);
      case 'lte':
        return a.lte(b);
    }
  } catch {
    throw new QuantError(
      'EVALUATION_FAILED',
      `values not orderable: ${typeof left} vs ${typeof right}`,
      {},
    );
  }
}

function resolveField(
  name: string,
  meta: StockMetaDto,
  asof: Date,
  snap: StockSnapshotDto | null,
): unknown {
  switch (name) {
    case 'code':
      return meta.code;
    case 'name':
      return meta.name;
    case 'industries':
      return meta.industries;
    case 'list_date':
      return parseIsoDate(meta.list_date);
    case 'float_pct':
      return meta.float_pct; // decimal-string; orderedCompare/scalarEq parse it
    case 'is_st':
      return isSt(meta.name);
    case 'exchange':
      return exchangeForCode(meta.code);
    case 'listed_days':
      return Math.floor((asof.getTime() - parseIsoDate(meta.list_date).getTime()) / 86_400_000);
    default:
      if (UNIVERSE_SNAPSHOT_FIELD_SET.has(name)) {
        return resolveSnapshotField(name, snap);
      }
      throw new QuantError('EVALUATION_FAILED', `unhandled universe field: ${name}`, {});
  }
}

/**
 * Snapshot-side field accessor. Returns the decimal-string value (the
 * `orderedCompare`/`scalarEq` helpers parse it via `Decimal`), or
 * `null` when the row's snapshot block is missing or the specific
 * field hasn't been populated yet.
 */
function resolveSnapshotField(name: string, snap: StockSnapshotDto | null): string | null {
  if (snap === null) return null;
  switch (name) {
    case 'price':
      return snap.price;
    case 'mkt_cap':
      return snap.derived.mkt_cap;
    case 'float_mkt_cap':
      return snap.derived.float_mkt_cap;
    case 'pe_ttm':
      return snap.derived.pe_ttm;
    case 'pe_dynamic':
      return snap.derived.pe_dynamic;
    case 'pb':
      return snap.derived.pb;
    case 'peg':
      return snap.derived.peg;
    case 'gross_margin_ttm':
      return snap.derived.gross_margin_ttm;
    case 'ret_1d':
      return snap.returns.ret_1d;
    case 'ret_5d':
      return snap.returns.ret_5d;
    case 'ret_10d':
      return snap.returns.ret_10d;
    case 'ret_20d':
      return snap.returns.ret_20d;
    case 'ret_90d':
      return snap.returns.ret_90d;
    case 'ret_250d':
      return snap.returns.ret_250d;
    case 'dde_main_net_inflow_3d':
      return snap.dde?.main_net_inflow_3d ?? null;
    case 'dde_main_net_inflow_5d':
      return snap.dde?.main_net_inflow_5d ?? null;
    case 'dde_main_net_inflow_10d':
      return snap.dde?.main_net_inflow_10d ?? null;
    case 'dde_main_net_inflow_20d':
      return snap.dde?.main_net_inflow_20d ?? null;
    case 'dde_main_inflow_ratio_3d':
      return snap.dde?.main_inflow_ratio_3d ?? null;
    case 'dde_main_inflow_ratio_5d':
      return snap.dde?.main_inflow_ratio_5d ?? null;
    case 'dde_main_inflow_ratio_10d':
      return snap.dde?.main_inflow_ratio_10d ?? null;
    case 'dde_main_inflow_ratio_20d':
      return snap.dde?.main_inflow_ratio_20d ?? null;
    default:
      return null;
  }
}

function isSt(name: string): boolean {
  const upper = name.trim().toUpperCase();
  return ST_PREFIXES.some((p) => upper.startsWith(p));
}

function exchangeForCode(code: string): string {
  if (!/^\d{6}$/.test(code)) return 'unknown';
  if (code.startsWith('920')) return 'bj';
  if (code.startsWith('60') || code.startsWith('68') || code.startsWith('900')) return 'sh';
  if (code.startsWith('00') || code.startsWith('30') || code.startsWith('20')) return 'sz';
  if (code.startsWith('4') || code.startsWith('8')) return 'bj';
  return 'unknown';
}

function parseIsoDate(iso: string): Date {
  // YYYY-MM-DD → UTC midnight.
  const y = Number.parseInt(iso.slice(0, 4), 10);
  const m = Number.parseInt(iso.slice(5, 7), 10);
  const d = Number.parseInt(iso.slice(8, 10), 10);
  return new Date(Date.UTC(y, m - 1, d));
}

function isIsoDateString(value: unknown): value is string {
  return typeof value === 'string' && /^\d{4}-\d{2}-\d{2}$/.test(value);
}

function coerceDate(value: unknown): Date | null {
  if (value instanceof Date) return value;
  if (isIsoDateString(value)) return parseIsoDate(value);
  return null;
}
